### PR TITLE
Fixed (PS-7187) - (Assertion failure: zipdecompress.cc:1432:!memcmp(F…

### PR DESCRIPTION
**Pull request by avodaniel**
…IL_PAGE_TYPE + page, FIL_PAGE_TYPE + page_zip->data, PAGE_HEADER - FIL_PAGE_TYPE))

The issue happens when one page is copied to another during
decompression and one page is encrypted and the other is not
or versions of encryption keys differ.
This patch introduces a solution when the versions
of encryption keys are compared. If they differ
then the whole page is copied including the header that
is not skipped.

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
@satya-bodapati could you take a look, please?
@hagrid-the-developer this PR has merge conflicts. Rebase most probably will help.

**Review by kamil-holubicki — CHANGES_REQUESTED** [review]
Would it be possible to rename the test by giving it a meaningful name?

File: storage/innobase/page/page0cur.cc [hunk 1]
```diff
@@ -1742,7 +1742,13 @@ rec_t *page_cur_insert_rec_zip(
       }
 
       /* Out of space: restore the page */
-      if (!page_zip_decompress(page_zip, page, FALSE)) {
+      const uint32_t encryption_key_version_in =
+          mach_read_from_4(page_zip->data + FIL_PAGE_ENCRYPTION_KEY_VERSION);
+      const uint32_t encryption_key_version_out =
+          mach_read_from_4(page + FIL_PAGE_ENCRYPTION_KEY_VERSION);
```
Comment (satya-bodapati) [inline storage/innobase/page/page0cur.cc:1748]: @hagrid-the-developer This feature (keyring key encryption)  is currently unused in PS and so FIL_PAGE_ENCRYPTION_KEY_VERSION  should never be changed.

Please find out how the KEY_VERSION changed when the feature is disabled.

File: mysql-test/suite/rocksdb/t/issue7187.test [hunk 2]
```diff
@@ -0,0 +1,389 @@
+INSTALL PLUGIN keyring_file SONAME 'keyring_file.so';
+
```
Comment (satya-bodapati) [inline mysql-test/suite/rocksdb/t/issue7187.test:2]: this should be in opt file along with keyring_file_data options

File: mysql-test/suite/rocksdb/t/issue7187.test [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+call mtr.add_suppression("Plugin keyring_file reported: 'File .*keyring' not found .*");
+call mtr.add_suppression("keyring_file initialization failure. Please check if the keyring_file_data points to readable keyring file or keyring file can be created in the specified location. The keyring_file will stay unusable until correct path to the keyring file gets provided");
+
+CREATE TABLESPACE tab02k_e ADD DATAFILE 'tab02k_e.ibd'  FILE_BLOCK_SIZE 02k ENCRYPTION='Y' ;
+CREATE TABLESPACE tab04k ADD DATAFILE 'tab04k.ibd'  FILE_BLOCK_SIZE 04k ENCRYPTION='N' ;
+CREATE TABLESPACE tab08k ADD DATAFILE 'tab08k.ibd'  FILE_BLOCK_SIZE 08k ENCRYPTION='N' ;
+CREATE TABLESPACE tab16k_e ADD DATAFILE 'tab16k_e.ibd'  FILE_BLOCK_SIZE 16k ENCRYPTION='Y' ;
```
Comment (satya-bodapati) [inline mysql-test/suite/rocksdb/t/issue7187.test:9]: how did this tablespace creation succeed? Keyring seems to be not initialized due to missing keyring_file_data.

File: mysql-test/suite/rocksdb/t/issue7187.test [hunk 4]
```diff
@@ -0,0 +1,389 @@
+INSTALL PLUGIN keyring_file SONAME 'keyring_file.so';
```
Comment (satya-bodapati) [inline mysql-test/suite/rocksdb/t/issue7187.test:1]: test should be in innodb suite with a proper name?

page_zip_decompress.test?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Comment by dutow** [comment]
Closing this, as it fixes an issue with encryption ids, which we removed as part of the keyring encryption removal.

**Pull request closed without merging**